### PR TITLE
General Blockchain Tests

### PR DESCRIPTION
**Pull request by winsvega**
example. WIP

description
https://github.com/ethereum/tests/issues/197

### CODE REVIEW

File: test/tools/libtesteth/TestHelper.cpp [hunk 1]
```diff
@@ -122,7 +122,19 @@ eth::Network stringToNetId(string const& _netname)
...
+	if (networks.size() == 0)
+	{
+		networks.push_back(eth::Network::FrontierTest);
+		networks.push_back(eth::Network::HomesteadTest);
+		networks.push_back(eth::Network::EIP150Test);
+		networks.push_back(eth::Network::EIP158Test);
+		networks.push_back(eth::Network::MetropolisTest);
+	}
+	return networks;
+}
```
Comment (pirapira) [inline test/tools/libtesteth/TestHelper.cpp]: Why not just a constant 
```
std::vector<eth::Network> constexpr networks {{
	eth::Network::FrontierTest,
	eth::Network::HomesteadTest,
	eth::Network::EIP150Test,
	eth::Network::EIP158Test,
	eth::Network::MetropolisTest
}};
```

File: test/tools/libtesteth/ImportTest.cpp [hunk 2]
```diff
@@ -535,8 +527,12 @@ void ImportTest::checkGeneralTestSectionSearch(json_spirit::mObject const& _expe
 		network.push_back(_network);
 
 	BOOST_CHECK_MESSAGE(network.size() > 0, TestOutputHelper::testName() + " Network array not set!");
-	vector<string> allowednetworks = {netIdToString(eth::Network::FrontierTest), netIdToString(eth::Network::HomesteadTest),
-				   netIdToString(eth::Network::EIP150Test), netIdToString(eth::Network::EIP158Test), netIdToString(eth::Network::MetropolisTest), "ALL"};
+	vector<string> allowednetworks;
+	vector<eth::Network> networks = test::getNetworks();
+	for (size_t i = 0; i < networks.size(); i++)
```
Comment (pirapira) [inline test/tools/libtesteth/ImportTest.cpp]: `for (auto const& net: networks)`

**Comment by hard-fault** [comment]
Hi,
I have a problem with building the code base. 
I followed all the steps given on the ethereum git home page.
But ran into issue with this command.

cmake --build .        # Build all default targets.

/Users/shreyasm/Desktop/Github/cpp-ethereum/build
bash-3.2$ cmake --build .
[  0%] Built target BuildInfo.h
[  1%] Performing download step (download, verify and extract) for 'boost'
-- verifying file...
       file='/Users/shreyasm/Desktop/Github/cpp-ethereum/deps/src/boost_1_63_0.tar.gz'
-- SHA256 hash of
    /Users/shreyasm/Desktop/Github/cpp-ethereum/deps/src/boost_1_63_0.tar.gz
  does not match expected value
    expected: 'fe34a4e119798e10b8cc9e565b3b0284e9fd3977ec8a1b19586ad1dec397088b'
      actual: 'efe2318e2797d664f9da43f38aa56440cd755b8ae81ba90edfdde4bc422d5711'
-- File already exists but hash mismatch. Removing...
-- Downloading...
   dst='/Users/shreyasm/Desktop/Github/cpp-ethereum/deps/src/boost_1_63_0.tar.gz'
   timeout='none'
-- Using src='https://github.com/ethereum/cpp-dependencies/releases/download/cache/boost_1_63_0.tar.gz'


It is stuck in that step for more than 24Hrs.

Please help me with this.

I am looking forward to fix some bugs.


Thanks,
Shreyas Badiger

**Comment by pirapira** [comment]
@hard-fault please file a new issue from https://github.com/ethereum/cpp-ethereum/issues/new

**Comment by pirapira** [comment]
Travis build is failing for Mac OS X.

**Comment by winsvega** [comment]
==> Downloading https://homebrew.bintray.com/bottles/llvm-4.0.1.sierra.bottle.tar.gz

Error: Failed to download resource "llvm"

Download failed: https://homebrew.bintray.com/bottles/llvm-4.0.1.sierra.bottle.tar.gz

**Comment by pirapira** [comment]
The failure needs to fixed before the PR is merged.

File: .travis.yml [hunk 3]
```diff
@@ -112,7 +112,7 @@ install:
     - ./scripts/install_deps.sh
 before_script:
     - ./scripts/prepare_ccache.sh
-    - ./scripts/build.sh $TRAVIS_BUILD_TYPE
+    - travis_wait 80 ./scripts/build.sh $TRAVIS_BUILD_TYPE
```
Comment (pirapira) [inline .travis.yml]: Sometimes, the above `brew install llvm` takes more than 10 minutes.  So `travis_wait` should also be there.

File: test/tools/jsontests/BlockChainTests.cpp [hunk 4]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
-				//Import Transactions
-				BOOST_REQUIRE(blObj.count("transactions"));
-				for (auto const& txObj: blObj["transactions"].get_array())
-				{
-					TestTransaction transaction(txObj.get_obj());
-					block.addTransaction(transaction);
-				}
+void fillBCTest(json_spirit::mObject& _o)
+{
+	string testname = TestOutputHelper::testName();
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: Do you need to copy the string of a reference is enough?

File: test/tools/jsontests/BlockChainTests.cpp [hunk 5]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
+		{
+			cnote << "Generating uncle block at test " << testname;
+			TestBlock uncle;
+			mObject uncleHeaderObj = uHObj.get_obj();
+			string uncleChainName = chainname;
+			if (uncleHeaderObj.count("chainname") > 0)
+				uncleChainName = uncleHeaderObj["chainname"].get_str();
+
+			overwriteUncleHeaderForTest(uncleHeaderObj, uncle, block.uncles(), *chainMap[uncleChainName]);
+			block.addUncle(uncle);
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: The variable `uncle` is not touched since declared.  This looks weird.

File: test/tools/jsontests/BlockChainTests.cpp [hunk 6]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
 
-			for (auto iterator = chainMap.begin(); iterator != chainMap.end(); iterator++)
-				delete iterator->second;
+		blObj["rlp"] = toHex(alterBlock.bytes(), 2, HexPrefix::Add);
+		blObj["blockHeader"] = writeBlockHeaderToJson(alterBlock.blockHeader());
 
-		}//fillin
-		else
+		mArray aUncleList;
+		for (size_t i = 0; i < alterBlock.uncles().size(); i++)
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: `for (auto const& uncle: alterBlock.uncles)`

File: test/tools/jsontests/BlockChainTests.cpp [hunk 7]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
-			TestBlockChain blockchain(genesisBlock);
+			if (blObj.count("expectException"))
+				BOOST_ERROR("Deprecated expectException field! " + testname);
 
-			if (o.count("genesisRLP") > 0)
+			blockchain.addBlock(alterBlock);
+			if (testChain.addBlock(alterBlock))
+				cnote << "The most recent best Block now is " <<  importBlockNumber << "in chain" << chainname << "at test " << testname;
+
+			if (test::Options::get().checkstate == true)
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: The 7 characters `== true` can be removed.

File: test/tools/jsontests/BlockChainTests.cpp [hunk 8]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
-					continue;
-				}
-				catch (...)
-				{
-					cnote << testname + "state sync or block import did throw an exception\n";
-					checkJsonSectionForInvalidBlock(blObj);
-					continue;
-				}
+				importedBlocks.push_back(alterBlock);
+				importedBlocks.at(importedBlocks.size()-1).clearState(); //close the state as it wont be needed. too many open states would lead to exception.
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: `importedBlocks.back().clearState();`

File: test/tools/jsontests/BlockChainTests.cpp (hunk 5 above)
Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: it is being altered inside overwriteUncleHeaderForTest

Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: OK.

File: test/tools/jsontests/BlockChainTests.cpp [hunk 9]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
 		//Select test by name if --singletest is set and not filling state tests as blockchain
 		if (!Options::get().fillchain && !TestOutputHelper::passTest(testname))
 		{
-			o.clear(); //don't add irrelevant tests to the final file when filling
-			continue;
+			if (_fillin)
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: `_fillin` and `Options::get().fillchain` look very similar.  Do they always have the same value?  If not, what is the difference between them?

File: test/tools/jsontests/BlockChainTests.cpp [hunk 10]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
+			{
+				o.clear(); //don't add irrelevant tests to the final file when filling
+				continue;
+			}
+			else
+			{
+				//Select test case from file by combining singleTestName and SingleTestNet
+				//singleTestName selects the fillerfile and SingleTestNet select particular test case in result test file
+				if (Options::get().singleTestName + "_" + Options::get().singleTestNet != testname)
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: This comparison is different from the comparison done in `TestOutputHelper::passTest(testname)`.  Do they serve different purposes?

Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: Does this comparison make sense when `singletestnet` option is not specified?

File: test/tools/jsontests/BlockChainTests.cpp (hunk 9 above)
Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: fillchain makes StateTests as Blockchain

File: test/tools/jsontests/BlockChainTests.cpp (hunk 10 above)
Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: the issue is that in blockchain filler test has  name  "test" 
but when blockchain test is generated it now has several tests with names
"test_EIP150" "test_Frontier"   and so on. 

so when you run filler you could select single test by  --singletest test
but when you run bctest you could selet single test by  --singletest test --singlenet Frontier

File: test/tools/jsontests/BlockChainTests.cpp [hunk 11]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
+					bool found = false;
+					for (auto& expect : expects)
 					{
-						ChainBranch::forceBlockchain(chainnetwork);
-						chainMap[chainname]->reset();
-						ChainBranch::resetBlockchain();
-						chainMap[chainname]->restoreFromHistory(importBlockNumber);
+						vector<string> netlist;
+						json_spirit::mObject& expectObj = expect.get_obj();
+						ImportTest::parseJsonStrValueIntoVector(expectObj["network"], netlist);
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: This should be done out of the for loop `for (auto& network : ...)`.

File: test/tools/jsontests/BlockChainTests.cpp (hunk 10 above)
Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: It wont run any test if singlenet is not defined

File: test/tools/jsontests/BlockChainTests.cpp [hunk 12]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
+						if (std::find(netlist.begin(), netlist.end(), test::netIdToString(network)) != netlist.end() ||
+							std::find(netlist.begin(), netlist.end(), "ALL") != netlist.end())
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: Do `const bool netliist_has_all = std::find(netlist.begin(), netlist.end(), "ALL");` out of the `for (auto& network : ...)` loop, and reuse the value.

File: test/tools/jsontests/BlockChainTests.cpp (hunk 11 above)
Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: why?

Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: There should be a check against incorrect network names in `netlist`.

File: test/tools/jsontests/BlockChainTests.cpp [hunk 13]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
-				else
-				{
-					ChainBranch::forceBlockchain(chainnetwork);
-					chainMap[chainname] = new ChainBranch(genesisBlock);
-					ChainBranch::resetBlockchain();
-				}
+				TestOutputHelper::setCurrentTestName(newtestname);
+				fillBCTest(jObj);
+				jObj["network"] = test::netIdToString(network);
+				tests[newtestname] = jObj;
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: There should be a check `tests[newtestname]` is not yet occupied.

File: test/tools/jsontests/BlockChainTests.cpp [hunk 14]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
+			testBCTest(o);
+		}
+	}
 
-				TestBlock block;
-				TestBlockChain& blockchain = chainMap[chainname]->blockchain;
-				vector<TestBlock>& importedBlocks = chainMap[chainname]->importedBlocks;
+	//Add generated tests to the result file
+	if (_fillin)
+	{
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: Can you assert `_v` is empty at this point?

File: test/tools/jsontests/BlockChainTests.cpp [hunk 15]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
-				//Import Transactions
-				BOOST_REQUIRE(blObj.count("transactions"));
-				for (auto const& txObj: blObj["transactions"].get_array())
-				{
-					TestTransaction transaction(txObj.get_obj());
-					block.addTransaction(transaction);
-				}
+void fillBCTest(json_spirit::mObject& _o)
+{
+	string const& testname = TestOutputHelper::testName();
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: The name of the variable should be `testName` with capital `N`.

File: test/tools/jsontests/BlockChainTests.cpp [hunk 16]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
+	BOOST_REQUIRE(_o.count("blocks"));
 
-				if (blObj.count("blockHeaderPremine"))
-				{
-					overwriteBlockHeaderForTest(blObj.at("blockHeaderPremine").get_obj(), block, *chainMap[chainname]);
-					blObj.erase("blockHeaderPremine");
-				}
+	mArray blArray;
+	size_t importBlockNumber = 0;
+	string chainname = "default";
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: `chainName`.

File: test/tools/jsontests/BlockChainTests.cpp [hunk 17]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
+	string chainnetwork = "default";
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: `chainNetwork`.

File: test/tools/libtesteth/TestHelper.cpp [hunk 18]
```diff
@@ -430,9 +442,12 @@ void executeTests(const string& _name, const string& _testPathAppendix, const st
 	if (Options::get().stats)
 		Listener::registerListener(Stats::get());
 
+	//Get the test name
 	string name = _name;
 	if (_name.rfind("Filler.json") != std::string::npos)
 		name = _name.substr(0, _name.rfind("Filler.json"));
+	else if (_name.rfind(".json") != std::string::npos)
+		name = _name.substr(0, _name.rfind(".json"));
```
Comment (pirapira) [inline test/tools/libtesteth/TestHelper.cpp]: Is this change related to this PR?

Comment (winsvega) [inline test/tools/libtesteth/TestHelper.cpp]: yes. finding bcForged tests

File: test/tools/jsontests/BlockChainTests.cpp (hunk 10 above)
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: That's a surprising behavior so there should at least be a warning.

File: test/tools/jsontests/BlockChainTests.cpp (hunk 11 above)
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: The reason is that the same thing should not be repeated.

File: test/tools/jsontests/BlockChainTests.cpp [hunk 19]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
-						blockFromFields.addUncle(uncle);
-					}
+	//make all values hex in pre section
+	State prestate(State::Null);
+	ImportTest::importState(_o["pre"].get_obj(), prestate);
+	_o["pre"] = fillJsonWithState(prestate);
 
-				checkBlocks(blockFromFields, blockFromRlp, testname);
+	for (auto iterator = chainMap.begin(); iterator != chainMap.end(); iterator++)
+		delete iterator->second;
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: For safety I would reset the pointer like  `iterator->second = nullptr` after `delete`.  Somebody might find an already `delete`d pointer.

File: test/tools/jsontests/BlockChainTests.cpp [hunk 20]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
+			cnote << testname + "state sync or block import did throw an exception\n";
+			checkJsonSectionForInvalidBlock(blObj);
+			continue;
+		}
+
+		//block from RLP successfully imported. now compare this rlp to test sections
+		BOOST_REQUIRE(blObj.count("blockHeader"));
+
+		//Check Provided Header against block in RLP
+		mObject emptyState;
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: If you can, add `const` here.

File: test/tools/jsontests/BlockChainTests.cpp [hunk 21]
```diff
@@ -117,304 +120,380 @@ void doBlockchainTestNoLog(json_spirit::mValue& _v, bool _fillin)
...
-	}//for tests
+
+		// ImportUncles
+		if (blObj["uncleHeaders"].type() != json_spirit::null_type)
+			for (auto const& uBlHeaderObj: blObj["uncleHeaders"].get_array())
+			{
+				mObject uBlH = uBlHeaderObj.get_obj();
+				BOOST_REQUIRE((uBlH.size() == 16));
+
+				mObject emptyState;
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: A variable of the same name already exists.  Maybe you can reuse the previously defined `emptyState`.

File: test/tools/libtesteth/TestHelper.cpp (hunk 18 above)
Comment (pirapira) [inline test/tools/libtesteth/TestHelper.cpp]: What is special about `bcForged`, and why do test names contain `.json`?

File: test/tools/jsontests/BlockChainTests.cpp (hunk 10 above)
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: Why do you use the same option for these two different purposes?  Why not `--testfilename`?  (Maybe this should go to a separate issue/PR).

File: test/tools/jsontests/BlockChainTests.cpp (hunk 9 above)
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: 1. what is generation mode?
2. can fillChain and _fillin occur at the same time?
3. can _fillin be false even when fillchain is true?

Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: 1. generating the test from source
2. yes
3. yes

File: test/tools/jsontests/BlockChainTests.cpp (hunk 10 above)
Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: cause --singletest run a single test from the file before
now we change things and in some cases filename = testname (but not for all test suites)

Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: lets discuss this and see if we could change it better

File: test/tools/libtesteth/TestHelper.cpp (hunk 18 above)
Comment (winsvega) [inline test/tools/libtesteth/TestHelper.cpp]: fillers for bcForged tests can not generate tests. I just keep the final .json test in src folder and instead filling, it is being copied to the destination.

File: test/tools/jsontests/BlockChainTests.cpp (hunk 11 above)
Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: I think this check then should be inside parseJsonStrValueIntoVector

File: test/tools/libtesteth/TestHelper.cpp (hunk 18 above)
Comment (pirapira) [inline test/tools/libtesteth/TestHelper.cpp]: Why do test names contain `".json"`?

File: test/tools/jsontests/BlockChainTests.cpp (hunk 9 above)
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: 1. `first with _fillin = true` does this suggest, `second time with _fillin == false`?
2. why cannot one select tests by testname, when this function is executed from statetests?  Can't we fill tests first and write the filled statetest to the disk, and then fill chain?

File: test/tools/jsontests/BlockChainTests.cpp (hunk 10 above)
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: I see.  I think `--singletest test_Frontier` is easier.

File: test/tools/jsontests/BlockChainTests.cpp (hunk 19 above)
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: Never mind.  `chainMap` is removed soon enough.

File: test/tools/jsontests/BlockChainTests.cpp (hunk 13 above)
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: OK, it was paranoid.

File: test/tools/jsontests/BlockChainTests.cpp (hunk 11 above)
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: I talked with @winsvega and the fix would complicate the code too much for this PR.

File: test/tools/jsontests/BlockChainTests.cpp (hunk 12 above)
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: Same as above.

File: test/tools/libtesteth/TestHelper.cpp (hunk 18 above)
Comment (pirapira) [inline test/tools/libtesteth/TestHelper.cpp]: I will try to separate test names and file names in a separate PR.

File: test/tools/libtesteth/ImportTest.cpp [hunk 22]
```diff
@@ -538,6 +538,27 @@ bool inArray(vector<T> const& _array, const T _val)
...
+	for (auto const& net : allnetworks)
+		allowedNetowks.push_back(test::netIdToString(net));
+
+	for (auto const& net: _networks)
+	{
+		if (!inArray(allowedNetowks, net))
+		{
+			//Can't use boost at this point
+			std::cerr << TestOutputHelper::testName() + " Specified Network not found: " + net << endl;
+			throw 1;
```
Comment (pirapira) [inline test/tools/libtesteth/ImportTest.cpp]: This causes an uncaught exception.  Maybe `exit(1)` is cleaner.
